### PR TITLE
(New PR) Added leaderboard functionally and unit tests

### PR DESCRIPTION
**Pull request by MO-W59**
### Targets Issue Leaderboard command #5

To avoid any further unnecessary notifications with actions taken to remedy the prior pull request I just created a new branch and properly pointed to dev as its upstream branch then applied changes too it. Afterwards I tested the functionally again and ran the unit tests again with working results.

Again, sorry for the confusion.

### Implements 

`leaderboard()` function in `game_commands.py`

`create_leaderboard()` method in `core.py`

`test_game_rejects_invalid_leaderboard_length()` in `test_game.py`

`test_game_rejects_invalid_leaderboard_sorting()` in `test_game.py`

`test_game_produces_valid_leaderboard()` in `test_game.py`

`test_request_leaderboard_no_register()` in `test_game_commands.py`

`test_request_leaderboard_no_game()` in `test_game_commands.py`

The default values for the leaderboard command are 'debt' and 5 therefore sending a bot message with the top 5 users based on debt, but users can input cranes and any other length. 

Achieved functionality:
```
>.leaderboard
1. user1 - 10,000 debt - 0 cranes
2. user2 - 8,989 debt - 2 cranes

>.leaderboard cranes
1. user2 - 8,989 debt - 2 cranes
2. user1 - 10,000 debt - 0 cranes

>.leaderboard debt 1
1. user1 - 10,000 debt - 0 cranes
```

Though with this solution it requires the user to enter a sorting type if they wish to enter a leaderboard length requirement.

#### test_game.py 

`test_game_rejects_invalid_leaderboard_length()` 
Checks proper rejection of a requested leaderboard length < 1.

`test_game_rejects_invalid_leaderboard_sorting()`
Checks proper rejection of an invalid sorting key, ie != "debt" or "cranes".

`test_game_produces_valid_leaderboard()` 
Checks proper leaderboard layout given 1 player and test values.

#### test_game_commands_cog.py 

`test_request_leaderboard_no_register()`
Checks command properly informs users if no one has registered and a leaderboard cannot be made.

`test_request_leaderboard_no_game()`
Checks command properly informs users when the game has not been properly loaded.

-Greg

### CODE REVIEW

File: src/word_debt_bot/game/core.py [hunk 1]
```diff
@@ -55,3 +55,25 @@ def add_debt(self, player_id: str, amount: int):
         state = self._state
         state[player_id].word_debt += amount
         self._state = state
+
+    def create_leaderboard(self, sort_by: str, lb_len: int):
+        if sort_by not in ["debt", "cranes"]:
```
Comment (Swiddis) [inline src/word_debt_bot/game/core.py:60]: Usability: It might be better UX if we `.lower()` here to avoid unnecessary errors, `.strip()` might also be helpful but I believe the command interface handles that automatically.

File: src/word_debt_bot/game/core.py [hunk 2]
```diff
@@ -55,3 +55,25 @@ def add_debt(self, player_id: str, amount: int):
...
+            raise ValueError("ordering is done by 'debt' or 'cranes'")
+        if lb_len < 1:
+            raise ValueError("requested leaderboard length must be 1 or greater")
+        # Make a sort key and sort users
+        if sort_by == "debt":
+            key = lambda u: u[1].word_debt
+        elif sort_by == "cranes":
+            key = lambda u: u[1].cranes
+        users = sorted(self._state.items(), key=key, reverse=True)
```
Comment (Swiddis) [inline src/word_debt_bot/game/core.py:69]: Nit: I think it'd be nicer to use `.values()` and to avoid indexing the tuple. For iteration you can also use [enumerate](https://docs.python.org/3/library/functions.html#enumerate) to avoid needing to manually keep track of the index when building the leaderboard string.

File: src/word_debt_bot/game/core.py [hunk 3]
```diff
@@ -55,3 +55,25 @@ def add_debt(self, player_id: str, amount: int):
...
+
+        # Produce leaderboard string to return
+        lb = ""
+        i = 1
+        for u in users:
+            lb += f"{i}. {u[1].display_name} - {u[1].word_debt:,} debt - {u[1].cranes:,} cranes\n"
```
Comment (Swiddis) [inline src/word_debt_bot/game/core.py:75]: Safety: Can we also add a guard against exceeding the 2000 character limit? Either by putting a limit on the number of users that can be requested at all, or by checking the length of the result before appending a new line to make sure we won't go over.

File: src/word_debt_bot/cogs/game_commands.py [hunk 4]
```diff
@@ -67,3 +67,17 @@ async def log(self, ctx, words: int):
             await ctx.send("Not registered! `.register`")
         except ValueError as err:
             await ctx.send(f"Error: {str(err)}")
+
+    @commands.command(name="leaderboard")
+    async def leaderboard(self, ctx, sort_by: str = "debt", lb_len: int = 5):
```
Comment (Swiddis) [inline src/word_debt_bot/cogs/game_commands.py:72]: Usability: I wonder if a page number would be better than a result count for displaying the leaderboard? If we specify a fixed size of each page, then we don't have to worry about exceeding the character limit (see below comment), and the user has more flexibility to go as far back in time as they want.

If we use [slice syntax](https://stackoverflow.com/a/509295/7543069) at `for u in users` we can easily get the range of users for a given page, without needing to use the `if i > lb_len` check.

**Review by Swiddis — CHANGES_REQUESTED** [review]
First of all, thanks for the contribution! 🎉

I like where you're going with it and appreciate a ton that you added some tests, the core of the functionality works great! I have some comments with regard to user experience, and think there's one more edge case to guard against. I think this is a great start though! 🎨✨😄

File: src/word_debt_bot/game/core.py (hunk 1 above)
Comment (MO-W59) [reply src/word_debt_bot/game/core.py:61]: `.lower()` added prior to the if branches for the sorting key.

Testing for extra whitespace from users showed that the command interface appears to be handling striping of it:
```
>.leaderboard      debt       10
1. User1 - 10,000 debt - 0 cranes
. . .
10. User10 ...
```

However, an input like: `.leaderboard   d ebt   3` causes a ValueError to be logged in discord.log

->Will work on a way to properly display error to user.

File: src/word_debt_bot/game/core.py (hunk 2 above)
Comment (MO-W59) [reply src/word_debt_bot/game/core.py]: Altered! That looks much nicer.

File: src/word_debt_bot/cogs/game_commands.py (hunk 4 above)
Comment (MO-W59) [reply src/word_debt_bot/cogs/game_commands.py:72]: If we took the maximum discord username length, plus the length of the static string characters per leaderboard entry, plus the maximum length of a `WordDebtPlayer`'s `word_debt` & `cranes` to calculate the maximum possible size of a leaderboard entry and divide 2,000 by that number we could produce a figure to represent the maximum entries available for any given page of the leaderboard:

According to [Discord] (https://support.discord.com/hc/en-us/articles/12620128861463-New-Usernames-Display-Names) max username length is 32.

`max_entries = 2000 // (max_user_name_len<32> + static_chars +  leaderboard_values)`

`//` in Python produces whole numbers only if I remember correctly, but I would need to properly test this math.

I'm guessing it would yield 30ish entries per page max, but it depends on the maximum number of digits for `word_debt` and `cranes`.

#### Alternative

Alternatively, we could add an entry to a leaderboard page until the requested amount of entries has been met and each time the current page string cannot accept a new entry before going over 2000 we would add the current page to the a leaderboard list of page strings and start a new page (repeating the process each time it would overflow 2000), finally iterate through that leaderboard list sending each page until all have been sent. (Also, would need to test this solution)

`leaderboard()` command could contain something like:

```python
for page in lb:
    await ctx.send(page)
```

`create_leaderboard()` `for` loop could contain something like:

```python
if len(page + entry) > 2000:
    lb.append(page)
    page =""
page += entry
```

The downside would be multiple messages from the bot (bot name & icon etc.) separating the actual leaderboard output and may look cluttered or unsightly.

Do you have a preference?

### Added slicing syntax
Added slicing and tuple enumeration, now reads, `for u, i in users[:lb_len]` and removed the unnecessary `if i > lb_len` looks nice and looks to be working well.

File: src/word_debt_bot/game/core.py (hunk 3 above)
Comment (MO-W59) [reply src/word_debt_bot/game/core.py]: added:
```python
if len(lb + entry) > 2000:
                break
```
But I am working on testing it's functionality.

File: src/word_debt_bot/cogs/game_commands.py (hunk 4 above)
Comment (MO-W59) [reply src/word_debt_bot/cogs/game_commands.py:72]: Could also simply limit leaderboard page size to 10.

File: src/word_debt_bot/game/core.py (hunk 3 above)
Comment (MO-W59) [reply src/word_debt_bot/game/core.py]: Ok, verified it works properly (was able to send 56 users with a request of 80).

File: src/word_debt_bot/cogs/game_commands.py (hunk 4 above)
Comment (Swiddis) [reply src/word_debt_bot/cogs/game_commands.py:72]: 10 might be good just because even if we hit the theoretical limit it might flood the channel to send 30 at once, 10 is what other bots do and it feels visually nice + is a clean number.

Comment (MO-W59) [reply src/word_debt_bot/cogs/game_commands.py]: Agreed, its looks pretty clean.

#### Sample inputs & outputs

```
> .leaderboard
1. User 1 ...
...
10. User 10 ...
```

```
> .leaderboard debt 2 
11. User11 ...
...
20. User20 ...
```

(order by cranes)
```
> .leaderboard cranes
1. User42 ...
...
10. User33 ...
```

(end of leaderboard)
```
> .leaderboard debt 5
41. User41 ...
42. User42 ...
```

(Request page that does not exists)
```
> .leaderboard
The leaderboard is not that long! Last page = 5
```

#### Picture example

Pic example

#### Code Changes

Changed the name of variables and arguments to better reflect a leaderboard page and requested page.
Changed name of `create_leaderboard()` to `get_leaderboard_page()`.
Added a check for when a user requests a page that does not exists. 
Added arithmetic to get the range of leaderboard entries for the requested page.
Added a check to only print to the end of the leaderboard if last page does not have 10 entries.
Updated unit tests to reflect changes.

However, this solution requires `import math`.

For error handling of invalid input syntax such as `.leaderboard d ebt 2` I posted a possible solution in Issue #13.

Cheers!,
Greg

**Review by Swiddis — APPROVED** [review]
